cpp/src/arrow/flight/test-integration-client.cc

```diff
@@ -60,23 +91,43 @@ int main(int argc, char** argv) {
     return -1;
   }
 
-  arrow::flight::Ticket ticket = info->endpoints()[0].ticket;
-  std::unique_ptr<arrow::RecordBatchReader> stream;
-  ABORT_NOT_OK(client->DoGet(ticket, schema, &stream));
-
-  std::shared_ptr<arrow::io::FileOutputStream> out_file;
-  ABORT_NOT_OK(arrow::io::FileOutputStream::Open(FLAGS_output, &out_file));
-  std::shared_ptr<arrow::ipc::RecordBatchWriter> writer;
-  ABORT_NOT_OK(arrow::ipc::RecordBatchFileWriter::Open(out_file.get(), schema, &writer));
-
-  std::shared_ptr<arrow::RecordBatch> chunk;
-  while (true) {
-    ABORT_NOT_OK(stream->ReadNext(&chunk));
-    if (chunk == nullptr) break;
-    ABORT_NOT_OK(writer->WriteRecordBatch(*chunk));
+  for (const arrow::flight::FlightEndpoint& endpoint : info->endpoints()) {
+    const auto& ticket = endpoint.ticket;
+
+    auto locations = endpoint.locations;
+    if (locations.size() == 0) {
+      locations = {arrow::flight::Location{FLAGS_host, FLAGS_port}};
+    }
+
+    for (const auto location : locations) {
+      std::cout << "Verifying location " << location.host << ':' << location.port
+                << std::endl;
+      // 3. Download the data from the server.
```

Comment:
You could factor out some of these canonical workflows into functions. e.g.

`ABORT_NOT_OK(ConsumeEndpoint(location, &retrieved_chunks))`